gcp.go

```diff
@@ -0,0 +1,209 @@
+	sanitizedLabels := sanitizeLabelsForGCP(labels)
+	log.Debugf("labels to add to PD volume: %s: %s", volumeID, sanitizedLabels)
+
+	project, location, name, err := parseVolumeID(volumeID)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+	disk, err := c.GetDisk(project, location, name)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+
+	// merge existing disk labels with new labels:
+	updatedLabels := make(map[string]string)
+	if disk.Labels != nil {
+		updatedLabels = maps.Clone(disk.Labels)
+	}
+	maps.Copy(updatedLabels, sanitizedLabels)
+	if maps.Equal(disk.Labels, updatedLabels) {
+		log.Debug("labels already set on PD")
+		return
+	}
+
+	req := &compute.ZoneSetLabelsRequest{
+		Labels:           updatedLabels,
+		LabelFingerprint: disk.LabelFingerprint,
+	}
+	op, err := c.SetDiskLabels(project, location, name, req)
+	if err != nil {
+		log.Errorf("failed to set labels on PD: %s", err)
+		promActionsTotal.With(prometheus.Labels{"status": "error", "storageclass": storageclass}).Inc()
+		return
+	}
+
+	waitForCompletion := func(_ context.Context) (bool, error) {
+		resp, err := c.GetGCEOp(project, location, op.Name)
+		if err != nil {
+			return false, fmt.Errorf("failed to set labels on PD %s: %s", disk.Name, err)
+		}
+		return resp.Status == "DONE", nil
+	}
+	if err := wait.PollUntilContextTimeout(context.TODO(),
+		time.Second,
+		time.Minute,
+		false,
+		waitForCompletion); err != nil {
+		log.Errorf("set label operation failed: %s", err)
+		return
+	}
+
+	log.Debug("successfully set labels on PD")
+	promActionsTotal.With(prometheus.Labels{"status": "success", "storageclass": storageclass}).Inc()
+}
+
+func deletePDVolumeLabels(c GCPClient, volumeID string, keys []string, storageclass string) {
+	if len(keys) == 0 {
+		return
+	}
+	sanitizedKeys := sanitizeKeysForGCP(keys)
+	log.Debugf("labels to delete from PD volume: %s: %s", volumeID, sanitizedKeys)
+
+	project, location, name, err := parseVolumeID(volumeID)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+	disk, err := c.GetDisk(project, location, name)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+	// if disk.Labels is nil, then there are no labels to delete
+	if disk.Labels == nil {
+		return
+	}
+
+	updatedLabels := maps.Clone(disk.Labels)
+	for _, k := range sanitizedKeys {
+		delete(updatedLabels, k)
+	}
+	if maps.Equal(disk.Labels, updatedLabels) {
+		return
+	}
+
+	req := &compute.ZoneSetLabelsRequest{
+		Labels:           updatedLabels,
+		LabelFingerprint: disk.LabelFingerprint,
+	}
+	op, err := c.SetDiskLabels(project, location, name, req)
+	if err != nil {
+		log.Errorf("failed to delete labels from PD: %s", err)
+		promActionsTotal.With(prometheus.Labels{"status": "error", "storageclass": storageclass}).Inc()
+		return
+	}
+
+	waitForCompletion := func(_ context.Context) (bool, error) {
+		resp, err := c.GetGCEOp(project, location, op.Name)
+		if err != nil {
+			return false, fmt.Errorf("failed to delete labels from PD %s: %s", disk.Name, err)
```

Comment:
Should probably increment the prometheus counter here.

```
promActionsTotal.With(prometheus.Labels{"status": "error", "storageclass": storageclass}).Inc()
```

Reply:
This closure may be called multiple times until success by the `PollUtilContext()` func a few lines lower.

If we only increment the metric after all retries fail (line 156) then we would have a single increment per disk attempt. I think that may be more desirable but I'm not positive.

I'd suggest only incrementing the counter after all retries have been exhausted. thoughts?

Reply:
Updating here and line 156 would result in a increasing the failure counter 2x for a single disk label update.

Reply:
ah right, I ready it as 2 separate cases but you are correct